File: android/app/src/main/java/com/kurt/bgmate/data/repository/RecommendRepositoryImpl.kt
package com.kurt.bgmate.data.repository

import android.content.Context
import android.util.Log
import com.kurt.bgmate.data.remote.llm.LlmClient
import com.kurt.bgmate.data.remote.llm.LlmMessage
import com.kurt.bgmate.data.remote.llm.LlmRequest
import com.kurt.bgmate.domain.model.BoardGame
import com.kurt.bgmate.domain.model.RecommendCondition
import com.kurt.bgmate.domain.model.RecommendResult
import com.kurt.bgmate.domain.repository.RecommendRepository
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okio.IOException
import org.json.JSONArray
import org.json.JSONObject
import javax.inject.Inject

class RecommendRepositoryImpl @Inject constructor(
    private val llmClient: LlmClient,
    @ApplicationContext private val context: Context
) : RecommendRepository {
    companion object {
        private const val TAG = "RecommendRepositoryImpl"
    }

    override suspend fun recommend(
        condition: RecommendCondition,
        includeNew: Boolean,
        ownedGames: List<BoardGame>
    ): List<RecommendResult> = withContext(Dispatchers.IO) {
        try {
            val prompt = buildPrompt(condition, ownedGames, includeNew)
            val request = LlmRequest(
                messages = listOf(LlmMessage(role = "user", content = prompt)),
                maxTokens = 8192
            )
            val response = llmClient.complete(request)
            parseRecommendResponse(response.text)
        } catch (e: Exception) {
            Log.e(TAG, "recommend: $e", e)
            throw IOException("추천을 가져오지 못했습니다.")
        }
    }

    private fun buildPrompt(
        condition: RecommendCondition,
        games: List<BoardGame>,
        includeNew: Boolean
    ): String {
        val fileName = "recommend_prompt_${if (includeNew) "all" else "owned"}.txt"

        val template = context.assets.open(fileName)
            .bufferedReader().readText()

        val moodText = condition.moods.joinToString(", ") { it.label }
        val gameListText = if (games.isEmpty()) "없음"
        else games.take(30).joinToString("\n") { "- ${it.name}" }

        return template
            .replace("{playerCount}", condition.playerCount.toString())
            .replace("{playTime}", condition.playTimeMinutes.label)
            .replace("{moods}", moodText)
            .replace("{gameList}", gameListText)
    }

    private fun parseRecommendResponse(raw: String): List<RecommendResult> {
        return try {
            val start = raw.indexOf('[')
            val end = raw.lastIndexOf(']') + 1
            if (start == -1 || end == 0) return emptyList()

            val array = JSONArray(raw.substring(start, end))
            (0 until array.length()).map { i ->
                val obj = array.getJSONObject(i)
                RecommendResult(
                    name = obj.getString("name"),
                    reason = obj.getString("reason"),
                    isOwned = obj.optBoolean("isOwned", false),
                    bggScore = obj.optDouble("bggScore", 0.0).toFloat(),
                    difficulty = obj.optString("difficulty", "")
                )
            }
        } catch (e: Exception) {
            Log.e(TAG, "parseRecommendResponse: $e", e)
            emptyList()
        }
    }
}

File: flutter/lib/data/repository/recommend_repository_impl.dart
import 'dart:convert';

import 'package:bgmate_flutter/data/remote/ai/llm_client.dart';
import 'package:bgmate_flutter/data/remote/ai/llm_request.dart';
import 'package:bgmate_flutter/domain/model/board_game.dart';
import 'package:bgmate_flutter/domain/model/recommend_condition.dart';
import 'package:bgmate_flutter/domain/model/recommend_result.dart';
import 'package:bgmate_flutter/domain/repository/game_repository.dart';
import 'package:bgmate_flutter/domain/repository/recommend_repository.dart';
import 'package:flutter/services.dart';

class RecommendRepositoryImpl implements RecommendRepository {
  final LlmClient llmClient;
  String? _systemPromptOwned;
  String? _systemPromptAll;

  RecommendRepositoryImpl({
    required this.llmClient,
    String? systemPromptOwned,
    String? systemPromptAll,
  })  : _systemPromptOwned = systemPromptOwned,
        _systemPromptAll = systemPromptAll {
    if (_systemPromptOwned == null || _systemPromptAll == null) {
      _loadPrompt();
    }
  }

  Future<void> _loadPrompt() async {
    _systemPromptOwned = await rootBundle.loadString(
      'assets/prompts/recommend_prompt_owned.txt',
    );
    _systemPromptAll = await rootBundle.loadString(
      'assets/prompts/recommend_prompt_all.txt',
    );
  }

  @override
  Future<List<RecommendResult>> recommend({
    required RecommendCondition condition,
    required bool includeNew,
    required List<BoardGame> ownedGames,
  }) async {
    final prompt = buildPrompt(condition, ownedGames, includeNew);
    final request = LlmRequest(
      messages: [LlmMessage(role: 'user', content: prompt)],
      maxTokens: 8192,
    );
    final response = await llmClient.complete(request);
    return parseRecommendResponse(response.text);
  }

  String buildPrompt(
    RecommendCondition condition,
    List<BoardGame> ownedGames,
    bool includeNew,
  ) {
    final moodText = condition.moods.map((e) => e.label).join(', ');
    final gameListText = ownedGames.isEmpty
        ? '없음'
        : ownedGames.take(30).map((e) => '- ${e.name}').join('\n');
    final template = includeNew ? _systemPromptAll : _systemPromptOwned;
    if (template == null) throw ArgumentError('Prompt is null');
    return template
        .replaceAll('{playerCount}', condition.playerCount.toString())
        .replaceAll('{playTime}', condition.playTimeMinutes.label)
        .replaceAll('{moods}', moodText)
        .replaceAll('{gameList}', gameListText);
  }

  static double? _parseScore(dynamic value) {
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value);
    return null;
  }

  List<RecommendResult> parseRecommendResponse(String raw) {
    try {
      final start = raw.indexOf('[');
      final end = raw.lastIndexOf(']') + 1;
      if (start == -1 || end == 0) return [];

      final jsonString = raw.substring(start, end);
      final List<dynamic> list = jsonDecode(jsonString);

      return list.map((item) {
        final map = item as Map<String, dynamic>;
        return RecommendResult(
          name: map['name'] as String,
          reason: map['reason'] as String,
          isOwned: map['isOwned'] as bool? ?? false,
          bggScore: _parseScore(map['bggScore']),
          difficulty: map['difficulty'] as String?,
        );
      }).toList();
    } catch (e) {
      return [];
    }
  }
}
